src/v/cluster/partition.cc

```diff
@@ -29,6 +29,40 @@ partition::partition(consensus_ptr r)
     } else if (_raft->log_config().is_collectable()) {
         _nop_stm = std::make_unique<raft::log_eviction_stm>(
           _raft.get(), clusterlog, _as);
+        if (config::shard_local_cfg().enable_idempotence.value()) {
+            _seq_stm = std::make_shared<raft::seq_stm>(
+              clusterlog, _raft.get(), config::shard_local_cfg());
+            _nop_stm->set_stm_manager(
+              std::static_pointer_cast<storage::stm_manager>(_seq_stm));
+            _raft->log().set_stm_manager(
+              std::static_pointer_cast<storage::stm_manager>(_seq_stm));
+        }
+    }
+}
+
+ss::future<result<raft::replicate_result>> partition::replicate(
+  model::record_batch_reader&& r, raft::replicate_options opts) {
+    return _raft->replicate(std::move(r), std::move(opts));
+}
+
+ss::future<checked<raft::replicate_result, kafka::error_code>>
+partition::replicate(
+  raft::batch_identity bid,
+  model::record_batch_reader&& r,
+  raft::replicate_options opts) {
+    if (bid.has_idempotent()) {
+        return _seq_stm->replicate(bid, std::move(r), std::move(opts));
+    } else {
+        return _raft->replicate(std::move(r), std::move(opts))
+          .then([](result<raft::replicate_result> result) {
+              if (result.has_value()) {
+                  return checked<raft::replicate_result, kafka::error_code>(
+                    result.value());
+              } else {
+                  return checked<raft::replicate_result, kafka::error_code>(
+                    kafka::error_code::unknown_server_error);
```

Comment:
i think we want to avoid using the kafka error codes outside the kafka layer.

Reply:
I think partition is part of the kafka layer, it owns seq_stm which validates requests and throws `out_of_order_sequence_number` when they don't pass. It feels weird to wrap kafka errors from down the stack to hide them from partition. Still want to change it?